### PR TITLE
autotest: Make ogr gpkg tests independent, parallel-safe

### PR DESCRIPTION
**Pull request by dbaston**
## What does this PR do?

Adds gdaltest.reopen helper
Makes ogr gpkg tests independent, parallel-safe

## What are related issues/pull requests?

#4407

## Tasklist

 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

File: autotest/ogr/ogr_gpkg.py [hunk 1]
```diff
@@ -7988,71 +8018,65 @@ def test_ogr_gpkg_arrow_stream_numpy_detailed_spatial_filter(tmp_vsimem, layer_t
...
+        gdal.OpenEx(
+            tmp_path / "read_only_test_ogr_gpkg_immutable/test.gpkg",
+            gdal.OF_VECTOR,
+            open_options=["IMMUTABLE=YES"],
+        )
+        is not None
+    )
+    assert gdal.GetLastErrorMsg() == ""
 
-    finally:
```
Comment (rouault) [inline autotest/ogr/ogr_gpkg.py:8052]: does pytest manage to remove the temporary directory despite our making it read-only ?

**Review by rouault — APPROVED** [review]

**Comment by rouault** [comment]
There seem to be a very minor coverage regression at https://coveralls.io/builds/62971018/source?filename=ogr%2Fogrsf_frmts%2Fgpkg%2Fogrgeopackagetablelayer.cpp#L4093  . So a situation where we would call CreateSpatialIndex() on a layer that has no geometry column.  That is triggered in master in test_ogr_gpkg_15() with ``ExecuteSQL("SELECT CreateSpatialIndex('non_spatial', '')")``

File: autotest/ogr/ogr_gpkg.py (hunk 1 above)
Comment (dbaston) [reply autotest/ogr/ogr_gpkg.py:8052]: Great question. Apparently yes, it does:
```
ls /tmp/pytest-of-dan/pytest-86/test_ogr_gpkg_immutable0/read_only_test_ogr_gpkg_immutable
# test.gpkg

pytest ogr/ogr_gpkg.py -k immutable
ls /tmp/pytest-of-dan/pytest-86/test_ogr_gpkg_immutable0/read_only_test_ogr_gpkg_immutable
# ls: cannot access '/tmp/pytest-of-dan/pytest-86/test_ogr_gpkg_immutable0/read_only_test_ogr_gpkg_immutable': No such file or directory
```

**Comment by dbaston** [comment]
> There seem to be a very minor coverage regression at https://coveralls.io/builds/62971018/source?filename=ogr%2Fogrsf_frmts%2Fgpkg%2Fogrgeopackagetablelayer.cpp#L4093 . So a situation where we would call CreateSpatialIndex() on a layer that has no geometry column. That is triggered in master in test_ogr_gpkg_15() with `ExecuteSQL("SELECT CreateSpatialIndex('non_spatial', '')")`

Restored: https://coveralls.io/builds/62973505/source?filename=ogr%2Fogrsf_frmts%2Fgpkg%2Fogrgeopackagetablelayer.cpp#L4093

**Pull request merged**